File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverse(ListNode* &head,int k,int n){
        ListNode *prev=NULL;
        ListNode *curr=head;
        ListNode *temp=head;
        int c=1;
        while(curr!=NULL && c<=k && n>0){
            temp=curr->next;
            curr->next=prev;
            prev=curr;
            curr=temp;
            c++;
        }
        n--;
        if(n==0){
            head->next=temp;
            return prev;
        }
        else{
            head->next=reverse(temp,k,n);
        }
        return prev;
    }
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode *temp=head;
        int count=0;
        while(temp!=NULL){
            count++;
            temp=temp->next;
        }
        ListNode *res=reverse(head,k,count/k);
        return res;
    }
};

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        if (head == null || k == 1) return head;
        
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode current = dummy, next = dummy, prev = dummy;
        
        int count = 0;
        while (current.next != null) {
            current = current.next;
            count++;
        }
        
        while (count >= k) {
            current = prev.next;
            next = current.next;
            for (int i = 1; i < k; i++) {
                current.next = next.next;
                next.next = prev.next;
                prev.next = next;
                next = current.next;
            }
            prev = current;
            count -= k;
        }
        
        return dummy.next;
    }
}
